### DISCUSSION TITLE
syntax problem

### DISCUSSION THREAD

**Discussion by A123-bansal123** [firedrakeproject/firedrake#2883] [Firedrake support]
Hello everyone

I tried to label the boundaries of a square domain with the syntax written below but CompliedSubDomain doesn't support in Firedrake, Please provide some alternative.

```

# Define the boundaries
left = CompiledSubDomain("near(x[0], 0)")
right = CompiledSubDomain("near(x[0], 1)")
bottom = CompiledSubDomain("near(x[1], 0)")
top = CompiledSubDomain("near(x[1], 1)")

# Define the boundary markers
boundary_markers = MeshFunction("size_t", mesh, mesh.topology().dim() - 1)
boundary_markers.set_all(0)
left.mark(boundary_markers, 1)
right.mark(boundary_markers, 2)
bottom.mark(boundary_markers, 3)
top.mark(boundary_markers, 4)

# Define the boundary conditions
u_left = Constant(0.0)
u_right = Constant(0.0)
u_bottom = Constant(1.0)
u_top = Constant(1.0)

bc_left = DirichletBC(V, u_left, left)
bc_right = DirichletBC(V, u_right, right)
bc_bottom = DirichletBC(V, u_bottom, bottom)
bc_top = DirichletBC(V, u_top, top)

bcs = [bc_left, bc_right, bc_bottom, bc_top]

```
One more thing, this doesn't work in Firedrake 
```
h = mesh.hmin()
```
and I use an alternative h = CellDiameter(mesh) , this create a long error. 
Please suggest some alternative syntax which I used in my code to work well.

**Comment by dham** [comment]
There are 3 ways to have boundary markers in Firedrake. The simplest one is that all the built-in meshes come with pre-marked boundaries. Just check the help of the built-in mesh function and it will tell you which edge has which marker. In this case, I guess you are probably using `UnitSquareMesh` for which the relevant help text is:

```
The boundary edges in this mesh are numbered as follows:

* 1: plane x == 0
* 2: plane x == 1
* 3: plane y == 0
* 4: plane y == 1
```

Second, if you read a mesh in from an external meshing program, you can usually mark edges in that program, and Firedrake will honour those markings.

Finally, you can mark labels on a mesh using the [`mark_entities`](https://www.firedrakeproject.org/firedrake.html#firedrake.mesh.AbstractMeshTopology.mark_entities) method of a Mesh, or create a new mesh with new markings from an existing mesh using [`RelabelledMesh`](https://www.firedrakeproject.org/firedrake.html#firedrake.mesh.RelabeledMesh)

If you want the mimimal cell diameter in a mesh, you can achieve this with:
```
project(CellDiameter(mesh), FunctionSpace(mesh, "DG", 0)).dat.data_ro.min()
```

**Comment by A123-bansal123** [comment]
When I define h = project(CellDiameter(mesh), FunctionSpace(mesh, "DG", 0)).dat.data_ro.min(), an error related to the installation is displayed.

error:
```
we suggest setting OMP_NUM_THREADS=1 to improve performance
Traceback (most recent call last):
  File "NS4.py", line 32, in <module>
    h = project(CellDiameter(M), FunctionSpace(mesh, "DG", 0)).dat.data_ro.min()
  File "PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/aparna/software/firedrake/src/firedrake/firedrake/functionspace.py", line 117, in FunctionSpace
    element = make_scalar_element(mesh, family, degree, vfamily, vdegree)
  File "PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/aparna/software/firedrake/src/firedrake/firedrake/functionspace.py", line 41, in make_scalar_element
    mesh.init()
AttributeError: module 'firedrake.mesh' has no attribute 'init'
```

Should I reinstall it?

**Comment by stephankramer** [comment]
This is not an installation error. I think you forgot to define `mesh` or named it something else than `mesh`. If you do `from firedrake import mesh`, `mesh` is a submodule of firedrake, not a mesh object - instead you need to pass in the actual mesh you want to use.

**Comment by A123-bansal123** [comment]
Thank you you are correct, there is some typo.
